tests/react/async.test.tsx

```diff
@@ -1143,3 +1144,229 @@ it('multiple derived atoms with dependency chaining and async write (#813)', asy
+      const count = get(countAtom)
+      await microtask()
+      console.log(`derived (${count})`)
+      const asyncCount = await get(asyncAtom)
+      console.log(`derived (${count})`, asyncCount)
+      setSelf(asyncCountAtom, asyncCount)
+    },
+    (_get, set, ...args: Parameters<Setter>) => set(...args)
+  )
+
+  const derivedSyncAtom = atom((get) => {
+    get(derivedAsyncAtom)
+  })
+
+  function useTest() {
+    const asyncCount = useAtomValue(asyncCountAtom)
+    useAtom(derivedSyncAtom)
+    const [count, setCount] = useAtom(countAtom)
+    return { count, setCount, asyncCount }
+  }
+  function TestComponent() {
+    const values = useTest()
+    return <div data-testid="test-comp">{JSON.stringify(values)}</div>
+  }
+  const store = createStore()
+
+  const Wrapper = ({ children }: React.PropsWithChildren) => (
+    <Provider store={store}>{children}</Provider>
+  )
+  render(<TestComponent />, { wrapper: Wrapper })
+
+  const setCount = (arg: SetStateAction<number>) => store.set(countAtom, arg)
+
+  await microtask()
+
+  expect(store.get(countAtom)).toBe(0)
+  expect(store.get(asyncCountAtom)).toBe(0)
+
+  await act(() => setCount(increment))
+  await act(() => setCount(increment))
+
+  resolve[0]!()
+  resolve[1]!()
+  resolve[2]!()
+
+  await macrotask()
+
+  expect(store.get(countAtom)).toBe(2)
+  expect(store.get(asyncCountAtom)).toBe(2)
+
+  await act(() => setCount(increment))
+  await act(() => setCount(increment))
+
+  resolve[3]!()
+  resolve[4]!()
+
+  await macrotask()
+
+  expect(store.get(countAtom)).toBe(4)
+  expect(store.get(asyncCountAtom)).toBe(4) // 3
+})
+
+// TODO: How do we unit test this code with store?
+it('[unit] resolves dependencies reliably after a delay', async () => {
+  const countAtom = atom(0)
+  const asyncCountAtom = atom(0)
+
+  const resolve: (() => void)[] = []
+  const asyncAtom = atom(async (get) => {
+    const count = get(countAtom)
+    await new Promise<void>((r) => resolve.push(r))
+    console.log(`resolved (${count})`)
+    return count
+  })
+
+  const derivedAsyncAtom = atom(
+    async (get, { setSelf }) => {
+      const count = get(countAtom)
+      await microtask()
+      console.log(`derived (${count})`)
+      const asyncCount = await get(asyncAtom)
+      console.log(`derived (${count})`, asyncCount)
+      setSelf(asyncCountAtom, asyncCount)
+    },
+    (_get, set, ...args: Parameters<Setter>) => set(...args)
+  )
+
+  const derivedSyncAtom = atom((get) => {
+    get(derivedAsyncAtom)
+  })
+
+  const store = createStore()
+  store.get(derivedSyncAtom)
+
+  expect(store.get(countAtom)).toBe(0)
+  expect(store.get(asyncCountAtom)).toBe(0)
+
+  store.set(countAtom, increment)
+  store.set(countAtom, increment)
+  await macrotask()
+
+  resolve[0]!()
+  // is it possible to do this test without react hooks?
+  resolve[1]!() // <--- Error: resolve[1] is not defined, it should be.
+  resolve[2]!()
+
+  await microtask()
+  await microtask()
+  await microtask()
```

Comment:
Do we need three of them because we have three atoms?

Reply:
I don't know.

In the other test, I arrived at this combination of micro / macro tasks experimentally. If I remove one, the expect doesn't see the update in time.

Reply:
I could be more general and use [waitFor](https://testing-library.com/docs/dom-testing-library/api-async/#waitfor), but I figured it would be better to be more explicit with this timing delay.

But let me know if you think this approach is too brittle.

Reply:
>  I figured it would be better to be more explicit with this timing delay.

Yeah, if the goal of test is about precise timing, it would be fine.
But, tests with `waitFor` can cover it, it sounds more reliable.

> But let me know if you think this approach is too brittle.

Yeah, I would like to see the `waitFor` version. If that works, it's probably better.

Reply:
ok, I switched to using `waitFor`.
And I got the unit test working with `store`.

So, right now both tests are failing, but if you add macrotask delays in places, they will pass.